base/broadcast.jl

```diff
@@ -623,8 +623,31 @@ eltypes(t::Tuple{Any}) = Tuple{_broadcast_getindex_eltype(t[1])}
 eltypes(t::Tuple{Any,Any}) = Tuple{_broadcast_getindex_eltype(t[1]), _broadcast_getindex_eltype(t[2])}
 eltypes(t::Tuple) = Tuple{_broadcast_getindex_eltype(t[1]), eltypes(tail(t)).types...}
 
+Base.@pure function promote_typejoin_union(::Type{T}) where T
```

Comment:
This is probably very wrong and @vtjnash will find lots of issues with this, but at least it should get the ball rolling.

Reply:
I can at least offer that this function doesn't seem to contain anything that has bitten me before.

Reply:
To follow that up, if anyone overloads `_promote_typejoin` (they might if they want to introduce a `Nothing` or `Missing` style of type that naturally forms a `Union` with other types - otherwise we have to consider these two as "blessed" types) or `typejoin` (they shouldn't do this) then you are in trouble here.

Reply:
@vtjnash Bump. I need your input to make progress on this.

Reply:
`promote_typejoin` is not pure (by design), so this is invalid

Reply:
Yes, but what would be a correct approach? :-)

Reply:
OK, I think I've found a correct approach: since `promote_typejoin` doesn't apply recursively inside tuples, we can split the part operation on tuple types into a separate method which is `@pure` since it only calls `typejoin`, and a non-pure part which is correctly inferred even without `@pure`.